File: typescript/tests/payment/domain/policies/paymentExpiration.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { paymentDeadlineReached, type PaymentDeadlineReached } from '@/payment/domain/events/paymentDeadlineReached.js';
import { Payment } from '@/payment/domain/payment/payment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { PaymentExpiration } from '@/payment/domain/policies/paymentExpiration.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';

describe('PaymentExpiration', () => {
  const policy = new PaymentExpiration();

  it('isDeadlineReached returns true when deadline passed', () => {
    const requestedAt = new Date('2026-06-01T00:00:00.000Z');
    const now = new Timestamp(new Date(requestedAt.getTime() + 48 * 3600 * 1000 + 1));
    const payment = paymentRequestedAt(requestedAt);

    expect(policy.isDeadlineReached(payment, now)).toBe(true);
  });

  it('isDeadlineReached returns false when deadline not passed', () => {
    const requestedAt = new Date('2026-06-01T00:00:00.000Z');
    const now = new Timestamp(new Date(requestedAt.getTime() + 47 * 3600 * 1000));
    const payment = paymentRequestedAt(requestedAt);

    expect(policy.isDeadlineReached(payment, now)).toBe(false);
  });

  it('isDeadlineReached returns true exactly at deadline', () => {
    const requestedAt = new Date('2026-06-01T00:00:00.000Z');
    const now = new Timestamp(new Date(requestedAt.getTime() + 48 * 3600 * 1000));
    const payment = paymentRequestedAt(requestedAt);

    expect(policy.isDeadlineReached(payment, now)).toBe(true);
  });

  it('isDeadlineReached rejects null payment', () => {
    expect(() => policy.isDeadlineReached(null as unknown as Payment, Timestamp.now())).toThrow();
  });

  it('isDeadlineReached rejects null now', () => {
    const payment = paymentRequestedAt(new Date());

    expect(() => policy.isDeadlineReached(payment, null as unknown as Timestamp)).toThrow();
  });

  it('evaluate returns expire payment command', () => {
    const payment = paymentRequestedAt(new Date());
    const event = paymentDeadlineReached(payment.id());

    const cmd = policy.evaluate(event);

    expect(cmd.aggregateId).toEqual(payment.id());
  });

  it('evaluate rejects null event', () => {
    expect(() => policy.evaluate(null as unknown as PaymentDeadlineReached)).toThrow();
  });

  function paymentRequestedAt(requestedAt: Date): Payment {
    const paymentId = generateId((value) => new PaymentId(value));
    return Payment.request(
      paymentId,
      new ClientReference(Uuid.fromString(crypto.randomUUID())),
      new Money(50),
      new Timestamp(requestedAt)
    );
  }
});

File: java/src/test/java/it/giannibombelli/wsc2026/payment/application/policies/PaymentExpirationTest.java
package it.giannibombelli.wsc2026.payment.application.policies;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.application.commands.ExpirePayment;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentDeadlineReached;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PaymentExpirationTest {

    private final PaymentExpiration policy = new PaymentExpiration();

    @Test
    void isDeadlineReached_returnsTrue_whenDeadlinePassed() {
        Instant requestedAt = Instant.parse("2026-06-01T00:00:00Z");
        Instant now = requestedAt.plusSeconds(48 * 3600 + 1);
        Payment payment = paymentRequestedAt(requestedAt);

        assertThat(policy.isDeadlineReached(payment, new Timestamp(now))).isTrue();
    }

    @Test
    void isDeadlineReached_returnsFalse_whenDeadlineNotPassed() {
        Instant requestedAt = Instant.parse("2026-06-01T00:00:00Z");
        Instant now = requestedAt.plusSeconds(47 * 3600);
        Payment payment = paymentRequestedAt(requestedAt);

        assertThat(policy.isDeadlineReached(payment, new Timestamp(now))).isFalse();
    }

    @Test
    void isDeadlineReached_returnsTrue_exactlyAtDeadline() {
        Instant requestedAt = Instant.parse("2026-06-01T00:00:00Z");
        Instant now = requestedAt.plusSeconds(48 * 3600);
        Payment payment = paymentRequestedAt(requestedAt);

        assertThat(policy.isDeadlineReached(payment, new Timestamp(now))).isTrue();
    }

    @Test
    void isDeadlineReached_rejectsNullPayment() {
        assertThatThrownBy(() -> policy.isDeadlineReached(null, Timestamp.now()))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void isDeadlineReached_rejectsNullNow() {
        Payment payment = paymentRequestedAt(Instant.now());

        assertThatThrownBy(() -> policy.isDeadlineReached(payment, null))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void evaluate_returnsExpirePaymentCommand() {
        Payment payment = paymentRequestedAt(Instant.now());
        PaymentDeadlineReached event = new PaymentDeadlineReached(payment.id());

        ExpirePayment cmd = policy.evaluate(event);

        assertThat(cmd.aggregateId()).isEqualTo(payment.id());
    }

    @Test
    void evaluate_rejectsNullEvent() {
        assertThatThrownBy(() -> policy.evaluate(null))
            .isInstanceOf(IllegalArgumentException.class);
    }

    private Payment paymentRequestedAt(Instant requestedAt) {
        PaymentId paymentId = EntityId.generate(PaymentId::new);
        return Payment.request(
            paymentId,
            new ClientReference(UUID.randomUUID()),
            new Money(new BigDecimal("50.00")),
            new Timestamp(requestedAt)
        );
    }
}
